src/Sentry/Internal/CacheDirectoryCoordinator.cs

```diff
@@ -0,0 +1,153 @@
+using Sentry.Extensibility;
+using Sentry.Internal.Extensions;
+
+namespace Sentry.Internal;
+
+internal class CacheDirectoryCoordinator : IDisposable
+{
+    private readonly IDiagnosticLogger? _logger;
+    private readonly IFileSystem _fileSystem;
+    private readonly object _gate = new();
+
+    private Stream? _lockStream;
+    private readonly string _lockFilePath;
+
+    private bool _acquired;
+    private bool _disposed;
+
+    public CacheDirectoryCoordinator(string cacheDir, IDiagnosticLogger? logger, IFileSystem fileSystem)
+    {
+        _logger = logger;
+        _fileSystem = fileSystem;
+        _lockFilePath = $"{cacheDir}.lock";
+
+        try
+        {
+            var baseDir = Path.GetDirectoryName(_lockFilePath);
+            if (!string.IsNullOrWhiteSpace(baseDir))
+            {
+                // Not normally necessary, but just in case
+                fileSystem.CreateDirectory(baseDir);
```

Comment:
thought: Accessing the file system in a `.ctor`

I'm a bit hesitant about IO in a `.ctor`.
Also, what should happen if an IO-Exception occurs?

Instead ... I suggest making the `.ctor` _private_, which just sets all fields without any IO-bound operations.
And have a static `Create` factory-method, with the same parameters of the `.ctor`, additionally doing the File-IO, and returning an instance of the `CacheDirectoryCoordinator`.

Reply:
I've moved the logic to the TryAcquire method... there was no particular reason to have it in the constructor.